spot-contracts/contracts/RolloverVault.sol

```diff
@@ -433,13 +427,15 @@ contract RolloverVault is
         // transfer underlying in
         underlying.safeTransferFrom(_msgSender(), address(this), underlyingAmtIn);
 
-        if (execRecover) {
-            // redeem tranches to maximize underlying balance for minting
-            recover();
-        }
-
         // tranche and mint perps as needed
-        _trancheAndMintPerps(perp, perpPrice, perpAmtOut + perpAmtToBurn);
+        {
+            uint256 perpBal = perp.balanceOf(address(this));
+            uint256 perpsReq = perpAmtOut + perpAmtToBurn;
+            if (perpBal < perpsReq) {
+                recover();
```

Comment:
Is it too inefficient to gate the recover op on whether there's enough underlying balance?

Gating tranching based on whether there's enough perp makes sense, but recover only *really* needs to happen if there's not enough underlying.

Perhaps, the lazy recover() could be moved into the tranche and mint function, where we already know the required values?